src/lib/field.ts

```diff
@@ -0,0 +1,1194 @@
+   */
+  mul(y: Field | bigint | number | string): Field {
+    if (this.isConstant() && isConstant(y)) {
+      return new Field(Fp.mul(this.toBigInt(), toFp(y)));
+    }
+    // if one of the factors is constant, return Scale AST node
+    if (isConstant(y)) {
+      let z = Snarky.field.scale(this.value, Field.#toConst(y));
+      return new Field(z);
+    }
+    if (this.isConstant()) {
+      let z = Snarky.field.scale(y.value, this.value[1]);
+      return new Field(z);
+    }
+    // create a new witness for z = x*y
+    let z = Snarky.existsVar(() =>
+      FieldConst.fromBigint(Fp.mul(this.toBigInt(), toFp(y)))
+    );
+    // add a multiplication constraint
+    Snarky.field.assertMul(this.value, y.value, z);
+    return new Field(z);
+  }
+
+  /**
+   * [Modular inverse](https://en.wikipedia.org/wiki/Modular_multiplicative_inverse) of this {@link Field} element. Equivalent to 1 divided by this {@link Field}, in the sense of modular arithmetic.
+   *
+   * @example
+   * ```ts
+   * const someField = Field(42);
+   * const inverse = someField.inv();
+   * inverse.assertEquals(Field(1).div(example)); // This statement is always true regardless of the value of `someField`
+   * ```
+   *
+   * **Warning**: This is a modular inverse. See {@link div} method for more details.
+   *
+   * @return A {@link Field} element that is equivalent to one divided by this element.
+   */
+  inv() {
+    if (this.isConstant()) {
+      let z = Fp.inverse(this.toBigInt());
+      if (z === undefined) throw Error('Field.inv(): Division by zero');
+      return new Field(z);
+    }
+    // create a witness for z = x^(-1)
+    let z = Snarky.existsVar(() => {
+      let z = Fp.inverse(this.toBigInt()) ?? 0n;
+      return FieldConst.fromBigint(z);
+    });
+    // constrain x * z === 1
+    Snarky.field.assertMul(this.value, z, FieldVar[1]);
+    return new Field(z);
+  }
+
+  /**
+   * Divide another "field-like" value through this {@link Field}.
+   *
+   * @example
+   * ```ts
+   * const x = Field(6);
+   * const quotient = x.div(Field(3));
+   *
+   * quotient.assertEquals(Field(2));
+   * ```
+   *
+   * **Warning**: This is a modular division in the pasta field. You can think this as the reverse operation of modular multiplication.
+   *
+   * @example
+   * ```ts
+   * const x = Field(2);
+   * const y = Field(5);
+   *
+   * const quotient = x.div(y);
+   *
+   * // If you try to print quotient - `console.log(quotient.toBigInt())` - you will realize that it prints a very big integer because this is a modular inverse.
+   * // You can use the reverse operation of division (multiplication) to prove the quotient is calculated correctly.
+   *
+   * quotient.mul(y).assertEquals(x);
+   * ```
+   *
+   * @param value - a "field-like" value to divide with the {@link Field}.
+   *
+   * @return A {@link Field} element equivalent to the modular division of the two value.
+   */
+  div(y: Field | bigint | number | string) {
```

Comment:
same here ^^ note on what errors are thrown and div-by-zero warnings as appropriate